paddle/fluid/platform/gpu_info.cc

```diff
@@ -227,72 +227,54 @@ void GpuMemoryUsage(size_t *available, size_t *total) {
                  error_code, CudaErrorWebsite());
 }
 
-size_t GpuMaxAllocSize() {
-  return std::max(GpuInitAllocSize(), GpuReallocSize());
-}
-
-size_t GpuInitAllocSize() {
-  if (FLAGS_initial_gpu_memory_in_mb > 0ul) {
-    // Initial memory will be allocated by FLAGS_initial_gpu_memory_in_mb
-    return static_cast<size_t>(FLAGS_initial_gpu_memory_in_mb << 20);
-  }
-
-  // FLAGS_initial_gpu_memory_in_mb is 0, initial memory will be allocated by
-  // fraction
+size_t GpuAvailableMemToAlloc() {
   size_t total = 0;
   size_t available = 0;
-
-  GpuMemoryUsage(&available, &total);
   size_t reserving = static_cast<size_t>(fraction_reserve_gpu_memory * total);
-
-  return static_cast<size_t>((total - reserving) *
-                             FLAGS_fraction_of_gpu_memory_to_use);
+  GpuMemoryUsage(&available, &total);
+  // If available less than minimum chunk size, no usable memory exists
+  size_t min_chunk_size = GpuMinChunkSize();
+  size_t available_to_alloc =
+      std::min(available > min_chunk_size ? available : 0, total - reserving);
+  VLOG(10) << "GPU usage " << (available >> 20) << "M/" << (total >> 20)
+           << "M, " << (available_to_alloc >> 20) << "M available to allocate";
+  return available_to_alloc;
 }
 
-size_t GpuReallocSize() {
-  if (FLAGS_reallocate_gpu_memory_in_mb > 0ul) {
-    // Additional memory will be allocated by
-    // FLAGS_reallocate_gpu_memory_in_mb
-    return static_cast<size_t>(FLAGS_reallocate_gpu_memory_in_mb << 20);
-  }
+size_t GpuMaxAllocSize() {
+  return std::max(GpuInitAllocSize(), GpuReallocSize());
+}
 
-  // FLAGS_reallocate_gpu_memory_in_mb is 0, additional memory will be
-  // allocated
-  // by fraction
-  size_t total = 0;
-  size_t available = 0;
+static size_t GpuAllocSize(bool realloc) {
+  size_t available_to_alloc = GpuAvailableMemToAlloc();
+  PADDLE_ENFORCE_GT(available_to_alloc, 0, "No enough available GPU memory");
+  // If FLAGS_initial_gpu_memory_in_mb is 0, then initial memory will be
+  // allocated by fraction
+  size_t flag_mb = realloc ? FLAGS_reallocate_gpu_memory_in_mb
+                           : FLAGS_initial_gpu_memory_in_mb;
+  size_t alloc_bytes =
+      (flag_mb > 0ul ? flag_mb << 20 : available_to_alloc *
+                                           FLAGS_fraction_of_gpu_memory_to_use);
+  PADDLE_ENFORCE_GT(available_to_alloc, alloc_bytes,
+                    "No enough available GPU memory");
+  VLOG(10) << "Alloc size is " << (alloc_bytes >> 20)
+           << " MiB, is it Re-alloc: " << realloc;
+  return alloc_bytes;
+}
 
-  GpuMemoryUsage(&available, &total);
-  size_t reserving = static_cast<size_t>(fraction_reserve_gpu_memory * total);
+size_t GpuInitAllocSize() { return GpuAllocSize(/* realloc = */ false); }
```

Comment:
Help reviewers: this is code clean up because I found the two has similar logic.